linkerd/lock/src/shared.rs

```diff
@@ -0,0 +1,166 @@
+        match self.try_acquire() {
+            Ok(Some(svc)) => Ok(Async::Ready(svc)),
+            Ok(None) => {
+                // Register the current task to be notified.
+                wait.register();
+                // Register the waiter's notify handle if one isn't already registered.
+                if let Some(notify) = wait.get_notify() {
+                    trace!("Registering waiter");
+                    self.waiters.push(notify);
+                }
+                debug_assert!(wait.is_waiting());
+                Ok(Async::NotReady)
+            }
+            Err(error) => Err(error),
+        }
+    }
+
+    pub fn release_and_notify(&mut self, value: T) {
+        trace!(waiters = self.waiters.len(), "Releasing");
+        debug_assert!(match self.state {
+            State::Claimed => true,
+            _ => false,
+        });
+        self.state = State::Unclaimed(value);
+        self.notify_next_waiter();
+    }
+
+    pub fn notify_next_waiter(&mut self) {
+        while let Some(waiter) = self.waiters.pop() {
+            if waiter.notify() {
+                trace!("Notified waiter");
+                return;
+            }
+        }
+    }
+
+    pub fn fail(&mut self, error: Arc<Error>) {
+        trace!(waiters = self.waiters.len(), %error, "Failing");
+        debug_assert!(match self.state {
+            State::Claimed => true,
+            _ => false,
+        });
+        self.state = State::Failed(error);
+
+        while let Some(waiter) = self.waiters.pop() {
+            waiter.notify();
+        }
+    }
+}
+
+mod waiter {
+    use futures::task::AtomicTask;
+    use std::sync::{Arc, Weak};
+
+    /// A handle held by Lock instances when waiting to be notified.
+    #[derive(Default)]
+    pub(crate) struct Wait(Arc<AtomicTask>);
+
+    /// A handle held by shared lock state to notify a waiting Lock.
+    ///
+    /// There may be at most one `Notify` instance per `Wait` instance at any one
+    /// time.
+    pub(super) struct Notify(Weak<AtomicTask>);
+
+    impl Wait {
+        /// If a `Notify` handle does not currently exist for this waiter, create
+        /// a new one.
+        pub(super) fn get_notify(&self) -> Option<Notify> {
+            if self.is_not_waiting() {
+                let n = Notify(Arc::downgrade(&self.0));
+                debug_assert!(self.is_waiting());
+                Some(n)
+            } else {
+                None
+            }
+        }
+
+        /// Register this waiter with the current task.
+        pub(super) fn register(&self) {
+            self.0.register();
+        }
+
+        /// Returns true iff there is currently a `Notify` handle for this waiter.
+        pub fn is_waiting(&self) -> bool {
+            Arc::weak_count(&self.0) == 1
+        }
+
+        /// Returns true iff there is not currently a `Notify` handle for this waiter.
+        pub fn is_not_waiting(&self) -> bool {
```

Comment:
Is a weak count every expected to be something other than 1 or 0? If so, we should comment somewhere so we remember. If not, is this function needed? It's just `!wait.is_waiting()`. Or the implementation could just be that, `!self.is_waiting()`.

Reply:
That's right. There may be at most one instance of `Notify` per `Wait` at any one time, as enforced by `get_notify`. I'll add `debug_assert`s to make this clearer.